### ISSUE TITLE
Speed difference bteween NETUSBEE.PRG and NETUSBEE.UCD

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#265]
This is basically an opposite of #263.

```
7d8 drivers + 7d8 netusbee.ucd: 300 kb/s
7d8 drivers + 7d8 netusbee.prg: 370-400 kb/s

a10 drivers + a10 netusbee.ucd: 300 kb/s
a10 drivers + a10 netusbee.prg: 330 kb/s (the speed decrease is reported as #264, let's ignore this one here)
```

7d8 = https://github.com/freemint/freemint/commit/7d850b71e432ca08941bf53806a5fd483f23adbe
a10 = https://github.com/freemint/freemint/commit/a10b393aa2fd55e856e7950308c1aa27fa12eb69

Tested on TT030 + ThunderStorm + Lightning + USB hub + USB stick. XFERRATE 1.2, Rwabs() numbers.

**Comment by mikrosk** [comment]
@DavidGZ , @Perdrix24 have you ever noticed this?

**Comment by mikrosk** [comment]
Can't say whether I confirm this on my MegaSTE: netus000.prg = 230 kb/s, netusbee.ucd = 220 kb/s. Can be just FreeMiNT overhead.

**Comment by DavidGZ** [comment]
> @DavidGZ , @Perdrix24 have you ever noticed this?

I don't remember but what I can say is that I've always accepted that TOS and MiNT drivers although share most of the code they have small differences that can make them behave different, for example the NetUSBee driver for MiNT uses a kernel thread for polling root hub status changes, also when a hub is connected another kernel thread is created.

This is why it is also a good idea to test both drivers when there are important changes in the source code.

**Comment by DavidGZ** [comment]
> Can't say whether I confirm this on my MegaSTE: netus000.prg = 230 kb/s, netusbee.ucd = 220 kb/s. Can be just FreeMiNT overhead.

Yes,I think this is the reason, we use some kernel facilities that can produce some overhead as you say

**Comment by DavidGZ** [comment]
> > Can't say whether I confirm this on my MegaSTE: netus000.prg = 230 kb/s, netusbee.ucd = 220 kb/s. Can be just FreeMiNT overhead.
> 
> Yes, I think this is the reason, we use some kernel facilities that can produce some overhead as you say.

Or the own kernel internal stuff (kernel scheduler, etc..)

**Comment by mikrosk** [comment]
Did a few more tests...

```
f030 + a10 + netusbee.prg: 310 kb/s
f030 + a10 + netusbee.ucd: 270 kb/s

ct60 + a10 + netusbee.prg: 470 kb/s
ct60 + a10 + netusbee.ucd: 460 kb/s
```

So you see, it seems that the 030 must be some kind of sweet spot. 000 is perhaps too slow to get a real speed difference, 060 perhaps too fast but don't you think @DavidGZ that 40 KB/s (Falcon030) and 70-100 KB/s (TT030) looks a bit more suspicious to attribute it only to some kernel overhead?

**Comment by DavidGZ** [comment]
> Did a few more tests...
> 
> ```
> f030 + a10 + netusbee.prg: 310 kb/s
> f030 + a10 + netusbee.ucd: 270 kb/s
> 
> ct60 + a10 + netusbee.prg: 470 kb/s
> ct60 + a10 + netusbee.ucd: 460 kb/s
> ```
> 
> So you see, it seems that the 030 must be some kind of sweet spot. 000 is perhaps too slow to get a real speed difference, 060 perhaps too fast but don't you think @DavidGZ that 40 KB/s (Falcon030) and 70-100 KB/s (TT030) looks a bit more suspicious to attribute it only to some kernel overhead?

I really don't know. It's strange that the difference because the kernel overhead doesn't appear with the Lightning, so probably you're right. But the drivers have their differences, I guess that the Lightning like the EtherNAT doesn't have to poll its root hub because they have the interrupt line connected. Polling in the MiNT driver is done by a kernel thread in the TOS driver is done in the USBTOOL.ACC by evnt_multi(), so by the AES. Although hubs are polled every second so I don't think this is enough to show those transfer rates disparities.......as I said I have no idea :-)

It could be interesting to see if you connect the USB stick without a hub (directly attached to the NetUSBee) has any effect in the results.

**Comment by mikrosk** [comment]
> It could be interesting to see if you connect the USB stick without a hub (directly attached to the NetUSBee) has any effect in the results.

Unfortunately it hasn't. :-/

**Comment by Perdrix24** [comment]
I don't have a FreeMiNT setup, I always test with TOS or MagiC. I rely on my friend Kroll to do tests on FreeMiNT, although I don't ask for speed tests. I know Roger Burrows once drastically increased the UCD speed. Perhaps one can find more optimizations in the Storage UDD. While I was working with HDDRIVER modules, once I let the USB drive be controlled by HDDRIVER through SCSIDRV, but the transfer rate was slightly slower. So maybe our Storage UDD is already optimized.

**Comment by mikrosk** [comment]
Even after correcting the [mistake with XFERRATE's PRG flags](https://github.com/freemint/freemint/issues/264#issuecomment-1174004571) the gap remains:
- XFERRATE in FreeMiNT A10: 380-370 KB/s
- XFERRATE in TOS with A10 drivers: 460 KB/s

**Comment by mikrosk** [comment]
So you see @Perdrix24 this can't be a STORAGE.UDD's fault as I'm using exactly the same UDD here with Lightning as well. So there must be something which NETUSBEE.PRG does differently from BLITZ030.PRG.

**Comment by Perdrix24** [comment]
I remember back in 2018, I looked at Blitz. I tried to use their assembler instead of our C. My notes were:

3. write/read_ptddata_to_fifo in assembler. Try 1st. Crashed big time. Brought over pack_fifo, unpack_fifo. Crashing, has to be Lightning-specific, can't use.

My assembler knowledge is near nil. Perhaps someone else can look at the differences between blitz and netusbee to make it faster.

**Comment by mikrosk** [comment]
@Perdrix24 but that part(s) is shared also with NETUSBEE.PRG, isn't it? So this can't be the difference which makes it so slow under FreeMiNT either.

**Comment by Perdrix24** [comment]
@mikrosk No, write_ptddata_to_fifo and read_ptddata_from_fifo are different. NetUSBee is in C, Blitz (vttusb) is in Assembler. I suspect they did it for performance.

Someone please look into this, and other differences there may be. Compare the files [isp116x-hcd.c](https://github.com/freemint/freemint/blob/master/sys/usb/src.km/ucd/vttusb/isp116x-hcd.c) and [isp116x-hcd.c](https://github.com/freemint/freemint/blob/master/sys/usb/src.km/ucd/netusbee/isp116x-hcd.c).

**Comment by mikrosk** [comment]
Actually I plan to do this. :) However what I meant is that yeah, sure, Blitz can be faster thanks to some of its optimisations but most likely this isn't the reason why its PRG and UCD variants provide similar performance? Because this is the problem here -- not NetUSBee driver slowness but only that its UCD flavour is significantly slower than PRG... (of course, there is always chance that vttusb's isp116x-hcd.c does have a special code path for UCD).

**Comment by mikrosk** [comment]
Whilte testing various setups on my Falcon, I managed to verify this one as well:

ct60@66 MHz:
netusbee.ucd: 460 kb/s
netusbee.prg: 470 kb/s

Falcon 030@16 MHz:
netusbee.ucd: 210-240 / 230 kb/s
netusbee.prg: 300 kb/s

Latest drivers, same hardware, same resolution. So the issue is still present.
